### PR TITLE
Explicit lazy parsing in SimpleFlatTableProducer

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

The T0 had a failure from the string parser which we have previously attributed to using the lazy option of the string parser. Switched code to allow lazy parsing to be set in the configuration. The default is no lazy parsing as turning on
lazy parsing can lead to problems.

#### PR validation:

Code compiles. Tested workflow 135.4 with the change and it ran fine.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by vlimant** [comment]
Hello, string parser are not only instantiated from where you indicated for nano, AFAICT. Should we change everywhere else ? maybe we can override the lazyness directly at the source (i.e. not leave the option of being lazy)

**Comment by Dr15Jones** [comment]
I took a different tack now. Now one sets `lazyEval` in the configuration explicitly if it is needed.

Personal note: the large number of `lazyEval = True` which were needed in my opinion points to the wrong class type (i.e. the base `reco::Candidate`) being read from the Event, instead of having modules which read the proper concrete type.

**Comment by vlimant** [comment]
my 2 cents, I don’t see how your suggested changes would help. you replaced the hard coded true, with configurable value that IMO will remain true throughout, and are unlikely to be changed.
the issue should probably be solved at the source with loading the necessary referenced object library  ; we have the input tag specified somewhere, so the fwk could know the type and pre-fetch (maybe this is not making sense ; just trying to find a practical solution)

**Comment by Dr15Jones** [comment]
>  I don’t see how your suggested changes would help.

The changes only help the exact problem seen in T0, where an instantiation of  the class `BXVectorSimpleFlatTableProducer<>` had the failure, i.e. SimpleTriggerL1EtSumFlatTableProducer with label l1EtSumTable.. In this change, that class never uses `lazyEval`.

To keep all the other modules functional required adding the `lazyEval`, where those setting now explicitly warn that those modules could have a threading problem.

**Comment by vlimant** [comment]
for what concerns : Personal note: the large number of lazyEval = True which were needed in my opinion points to the wrong class type (i.e. the base reco::Candidate) being read from the Event, instead of having modules which read the proper concrete type.

indeed, many collections, if not the majority are accepted as  `edm::View<reco::Candidate>` so that we do not have to declare a million module types to each and every possible class deriving from `reco::Candidate`. Maybe we can fix this issue this way ; by never accessing anything as view but direct type, and declare as many ed modules as necessary

**Comment by Dr15Jones** [comment]
> Maybe we can fix this issue this way ; by never accessing anything as view but direct type, and declare as many ed modules as necessary. That is IF we cannot fix this at its root

That is the surest path forward. Also, using the lazy evaluation has a performance cost as it must constantly do a dynamic_cast to check each object and it might also be doing string comparisons as well.

**Comment by makortel** [comment]
> > I see. do you have the GH issue number (or else) reporting that issue? I must have missed it.
> 
> All I have is this discussion: https://cms-talk.web.cern.ch/t/paused-job-in-promptreco-for-minimumbias-pd-run-378378/37744/11

There is also an issue https://github.com/cms-sw/cmssw/issues/33084

**Comment by Dr15Jones** [comment]
please test

**Comment by vlimant** [comment]
Can we maybe try to proposed patch in 14.0.X and check when 14.0 gets deployed if we still experience crashes of the sort at T0.
I'd rather have the lazy=true by default (unspecified) at the module configuration level (not at the function/expression configuration level) and set the false where it's known we'll use the specified and nothing inheriting from it (https://github.com/cms-sw/cmssw/commit/03f43a32a433e0282a33e514c071780e8d726077)

If we cannot fix it at ROOT level, then we'll do the big change of not using view systematically, but the exact type systematically.

**Comment by Dr15Jones** [comment]
> I'd rather have the lazy=true by default (unspecified) at the module configuration level (not at the function/expression configuration level) and set the false where it's known we'll use the specified and nothing inheriting from it 

I disagree, but please feel free to use this code as a basis for your own changes.

**Comment by vlimant** [comment]
on third thought, indeed, having it at the branch level is more appropriate, although still not perfect.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

I ran all the failing workflows locally and after this round of changes, they all ran.

**Comment by vlimant** [comment]
enable nano

**Comment by vlimant** [comment]
please test

**Comment by vlimant** [comment]
+1

**Comment by antoniovilela** [comment]
> I took a different tack now. Now one sets `lazyEval` in the configuration explicitly if it is needed.
> 
> Personal note: the large number of `lazyEval = True` which were needed in my opinion points to the wrong class type (i.e. the base `reco::Candidate`) being read from the Event, instead of having modules which read the proper concrete type.

Should an issue be opened to address this?

**Comment by antoniovilela** [comment]
-orp

**Comment by antoniovilela** [comment]
Need to fix the conflict.

**Comment by antoniovilela** [comment]
@Dr15Jones 
Could you quickly fix the conflict?
Many thanks.

**Comment by Dr15Jones** [comment]
enable nano

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
enable nano

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by antoniovilela** [comment]
please test
   - Should now get rid of the RelVal errors.

**Comment by antoniovilela** [comment]
urgent
   - To get this in before other Nano updates, this was already rebased twice.

**Comment by AdrianoDee** [comment]
please test with https://github.com/cms-data/RecoBTag-Combined/pull/57
(Didn't want to chime in but I was tagged here by the bot. I think it is either this either we wait for the next IB?)

**Comment by Dr15Jones** [comment]
please test with https://github.com/cms-data/RecoBTag-Combined/pull/57

File: PhysicsTools/NanoAOD/python/boostedTaus_cff.py [hunk 1]
```diff
@@ -11,7 +11,8 @@
 
 finalBoostedTaus = cms.EDFilter("PATTauRefSelector",
     src = cms.InputTag("slimmedTausBoosted"),
-    cut = cms.string("pt > 40 && tauID('decayModeFindingNewDMs') && (tauID('byVVLooseIsolationMVArun2DBoldDMwLT') || tauID('byVVLooseIsolationMVArun2DBnewDMwLT'))")
+    cut = cms.string("pt > 40 && tauID('decayModeFindingNewDMs') && (tauID('byVVLooseIsolationMVArun2DBoldDMwLT') || tauID('byVVLooseIsolationMVArun2DBnewDMwLT'))"),
+    lazyzEval = cms.untracked.bool(True)
```
Comment (hqucms) [inline PhysicsTools/NanoAOD/python/boostedTaus_cff.py:15]: typo: `lazyzEval` -> `lazyEval`
But actually is this implemented for `PATTauRefSelector`? I don't seem to find it anywhere...

File: PhysicsTools/NanoAOD/python/leptonTimeLifeInfo_common_cff.py [hunk 2]
```diff
@@ -48,12 +48,12 @@
...
-    refitSVx = Var("?hasSV()?sv().x():0", float, doc="x coordinate of SV", precision=10),
-    refitSVy = Var("?hasSV()?sv().y():0", float, doc="y coordinate of SV", precision=10),
-    refitSVz = Var("?hasSV()?sv().z():0", float, doc="z coordinate of SV", precision=10),
-    refitSVchi2 = Var("?hasSV()?sv().normalizedChi2():0", float, doc="reduced chi2, i.e. chi2/ndof, of SV fit", precision=8),
-    #refitSVndof = Var("?hasSV()?sv().ndof():0", float, doc="ndof of SV fit", precision=8),#MB: not important
+    hasRefitSV = Var("hasSV()", bool, doc="has SV refit using miniAOD quantities", lazyEval=True),
+    refitSVx = Var("?hasSV()?sv().x():0", float, doc="x coordinate of SV", precision=10, lazyEval=True),
+    refitSVy = Var("?hasSV()?sv().y():0", float, doc="y coordinate of SV", precision=10, lazyEval=True),
+    refitSVz = Var("?hasSV()?sv().z():0", float, doc="z coordinate of SV", precision=10, lazyEval=True),
+    refitSVchi2 = Var("?hasSV()?sv().chi2():0", float, doc="chi2 of SV fit", precision=8, lazyEval=True),
```
Comment (hqucms) [inline PhysicsTools/NanoAOD/python/leptonTimeLifeInfo_common_cff.py:55]:
```suggestion
    refitSVchi2 = Var("?hasSV()?sv().normalizedChi2():0", float, doc="reduced chi2, i.e. chi2/ndof, of SV fit", precision=8, lazyEval=True),
```
I think this is causing the difference in NANO.

**Comment by hqucms** [comment]
Do we understand why there are so many differences in these two WFs?

WF #	Differences
[/11634.0_TTbar_14TeV+2021/all_mini_RECO_step3_inMINIAODSIM](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-17-1100+340323/62393/validateJR//11634.0_TTbar_14TeV+2021/all_mini_RECO_step3_inMINIAODSIM)	1362
[/11634.0_TTbar_14TeV+2021/all_RECO_step3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-04-17-1100+340323/62393/validateJR//11634.0_TTbar_14TeV+2021/all_RECO_step3)	1573

File: PhysicsTools/NanoAOD/python/boostedTaus_cff.py (hunk 1 above)
Comment (Dr15Jones) [reply PhysicsTools/NanoAOD/python/boostedTaus_cff.py]: fixed. Thanks!

File: PhysicsTools/NanoAOD/python/leptonTimeLifeInfo_common_cff.py (hunk 2 above)
Comment (Dr15Jones) [reply PhysicsTools/NanoAOD/python/leptonTimeLifeInfo_common_cff.py]: fixed. Thanks again!

**Comment by Dr15Jones** [comment]
please test

**Comment by hqucms** [comment]
We still get a large number of differences in wf 11634.0 -- anything wrong in the IB?

**Comment by Dr15Jones** [comment]
> We still get a large number of differences in wf 11634.0 -- anything wrong in the IB?

It seems that other PRs are also showing large differences in 11634.0: see

https://github.com/cms-sw/cmssw/pull/44624#issuecomment-2064052844

**Comment by makortel** [comment]
> > We still get a large number of differences in wf 11634.0 -- anything wrong in the IB?
> 
> It seems that other PRs are also showing large differences in 11634.0: see
> 
> [#44624 (comment)](https://github.com/cms-sw/cmssw/pull/44624#issuecomment-2064052844)

Now also in an issue

**Comment by hqucms** [comment]
@Dr15Jones I think this can be superseded by https://github.com/cms-sw/cmssw/pull/44782 -- let me know what you think.

**Comment by Dr15Jones** [comment]
@hqucms wrote:
> I think this can be superseded by https://github.com/cms-sw/cmssw/pull/44782 -- let me know what you think.

I agree that using the concrete types along with the ability to explicitly turn on lazy evaluation from the configuration is a better approach. Once #44782 is merged, we can close this PR.

**Comment by antoniovilela** [comment]
https://github.com/cms-sw/cmssw/pull/44782 is merged.

**Comment by hqucms** [comment]
please close

superseded by https://github.com/cms-sw/cmssw/pull/44782

**Pull request closed without merging**